cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatusTool.java

```diff
@@ -0,0 +1,379 @@
+    bw.write(masterShutdownTime.toString());
+    bw.newLine();
+    //Write checksum
+    bw.write(masterChecksum);
+    bw.newLine();
+
+    //Write region map
+    for (Map.Entry<String, Long> entry: map.entrySet()) {
+      bw.write(entry.getKey() + delimiter + entry.getValue());
+      bw.newLine();
+    }
+    bw.close();
+  }
+
+  private static Map<String, Long> getMapFromTable(String qualifier) throws IOException {
+    HBaseTableUtil tableUtil = new HBaseTableUtilFactory(cConf).get();
+    HTable hTable = tableUtil.createHTable(hConf, getReplicationStateTableId(tableUtil));
+
+    // Scan the table to scan for all regions.
+    ScanBuilder scan = tableUtil.buildScan();
+    scan.addColumn(Bytes.toBytes(ReplicationConstants.ReplicationStatusTool.TIME),
+                   Bytes.toBytes(qualifier));
+    Result result;
+    HashMap<String, Long> timeMap = new HashMap<>();
+
+    try (ResultScanner resultScanner = hTable.getScanner(scan.build())) {
+      while ((result = resultScanner.next()) != null) {
+        ByteBuffer row = ByteBuffer.wrap(result.getRow());
+        getNextPart(row); // Ignore cluster ID;
+        String region = new String(getNextPart(row), "UTF-8");
+        long timestamp = getTimeFromResult(result, qualifier);
+
+        if (timeMap.get(region) == null || timestamp > timeMap.get(region)) {
+          timeMap.put(region, timestamp);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Error while reading table:" + e);
+      throw Throwables.propagate(e);
+    } finally {
+      hTable.close();
+    }
+    return timeMap;
+  }
+
+  private static byte[] getNextPart(ByteBuffer byteBuffer) {
+    int len = byteBuffer.getInt();
+    if (byteBuffer.remaining() < len) {
+      throw new BufferUnderflowException();
+    }
+    byte[] bytes = new byte[len];
+    byteBuffer.get(bytes, 0, len);
+    return bytes;
+  }
+
+  private static TableId getReplicationStateTableId(HBaseTableUtil tableUtil) throws IOException {
+    String tableName = hConf.get(ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_NAME);
+    String ns = hConf.get(ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_NAMESPACE);
+    TableId tableId = tableUtil.createHTableId(
+      (ns != null)
+        ? new NamespaceId(ns)
+        : new NamespaceId(ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_DEFAULT_NAMESPACE),
+      (tableName != null)
+        ? tableName
+        : ReplicationConstants.ReplicationStatusTool.REPLICATION_STATE_TABLE_DEFAULT_NAME);
+    return tableId;
+  }
+
+  private static long getTimeFromResult(Result result, String qualifier) {
+    long timestamp = 0L;
+    List<Cell> timeVal = result.getColumnCells(Bytes.toBytes(ReplicationConstants.ReplicationStatusTool.TIME),
+                                               Bytes.toBytes(qualifier));
+    if (timeVal.size() > 0) {
+      ByteBuffer timeBuffer =  ByteBuffer.wrap(CellUtil.cloneValue(timeVal.get(0)));
```

Comment:
remove extra space before ByteBuffer. You can also use result.getValueAsByteBuffer() instead of getting the list of cells, cloning the first, and wrapping it.